File: android/src/main/java/org/ligi/intentchoice/utils/IntentDescriber.java
package org.ligi.intentchoice.utils;

import android.content.Intent;

import com.google.common.base.Joiner;

public class IntentDescriber {
    private final Intent intent;

    public IntentDescriber(Intent intent) {
        this.intent = intent;
    }

    public String getUserFacingCategoriesString() {
        final String categories=Joiner.on(" ").join(intent.getCategories());
        return categories.replaceAll("android\\.intent\\.category\\.","");
    }

    public String getClaspedCategoriesString() {
        if (intent.getCategories()==null) {
            return "";
        }

        if (intent.getCategories().size()>1) {
            return "("+getUserFacingCategoriesString()+")";
        } else {
            return getUserFacingCategoriesString();
        }
    }

    public String getUserFacingActionString() {
        return intent.getAction().replace("android.intent.action.", "");
    }

    public String getUserFacingIntentDescription() {
        final StringBuilder res = new StringBuilder(getUserFacingActionString());
        if (intent.getCategories()!=null) {
            res.append("@" + getClaspedCategoriesString());
        }
        if (intent.getData()!=null) {
            res.append(intent.getData());
        }
        return res.toString();
    }

}

File: android/src/main/java/org/ligi/intento/utils/IntentDescriber.kt
package org.ligi.intento.utils

import org.ligi.intento.model.SimpleIntent

class IntentDescriber(val intent: SimpleIntent) {

    val userFacingCategoriesString: String
        get() {
            val categories = intent.categories.joinToString(separator = " ")
            return categories.replace("android\\.intent\\.category\\.".toRegex(), "")
        }

    val claspedCategoriesString: String
        get() {
            if (intent.categories == null) {
                return ""
            }

            if (intent.categories.size > 1) {
                return "($userFacingCategoriesString)"
            } else {
                return userFacingCategoriesString
            }
        }

    val userFacingActionString: String
        get() = intent.action.replace("android.intent.action.", "")

    val userFacingIntentDescription: String
        get() {
            val res = StringBuilder(userFacingActionString)
            if (!intent.categories.isEmpty()) {
                res.append("@$claspedCategoriesString")
            }
            if (intent.data != null) {
                res.append(":${intent.data}")
            }
            return res.toString()
        }

}
